src/pip/_internal/cli/base_command.py

```diff
@@ -186,6 +186,12 @@ def _main(self, args):
                 "This will become an error in pip 21.0."
             )
 
+        runner = self.run if 'pdb' in sys.modules else self._trap_run
```

Comment:
Ouch, this feels so terribly hacky. Is there no nother way to detect `python -m pdb`?

Also, does this also cover alternative pdb frontends e.g. ipdb, pudb?

Reply:
Very hacky indeed, and maybe even dangerous. I'm unsure if there are environments (Juypter comes to mind) where pdb might have been imported prior to invoking pip. Of course, pip only supports invocation from the command-line, so such a case would likely be unsupported.

I searched around and didn't find anything better. I looked at the pdb code and it doesn't set any state. With Python 3.8, construction of the `pdb.Pdb` class does raise a `sys.audit` event, but best I can tell, one can only add hooks for these events, not detect them after the fact.

Perhaps one could use the garbage collector to detect the presence of `Pdb` instances. That approach would avoid cases where `pdb` was imported but not invoked. Would you like me to draft that approach?

Or maybe it would be better to have an explict entry point, where `python -m pip.debug` or `-m pip.direct` would bypass the error trapping. Such an approach would have the unfortunate restriction of being more difficult to discover. Ideally, the presence of `-m pdb` should be enough to enable debugging.

I could also add an escape hatch, an environment variable like PIP_TRAP_ERRORS, which if true would use the trapping behavior, if false would use the direct behavior, but would fall back to the inference above (trap errors except when pdb is detected, however that happens).

My instinct, however, is that this escape hatch or more sophisticated detection behavior is unlikely to be needed, so I'd propose to go with this simple approach for now.

> Does this cover alternative pdb frontends?

I checked IPython, and it does import pdb as a matter of course, even when the debugger isn't engaged:

```
~ $ pip-run -q ipython -- -c "import sys, IPython; print('pdb' in sys.modules)"
True
```

So that's at least one case where the pdb detection would return a false positive.

That appears not to be the case for pudb:

```
pip-run --use-pep517 -q pudb -- -c "import sys, pudb; print('pdb' in sys.modules)"
False
```

Though I did find places where pudb imports `bdb`, suggesting that instances of `bdb.Bdb` might be the most accurate detection mechanism.

Reply:
In 501258d, I implement the `any(isinstance(ob, bdb.Bdb))` approach.

Reply:
@jaraco have you tried to check this via [`gettrace()`](https://docs.python.org/3/library/sys.html#sys.gettrace)? If it `is not None` then you're running under some kind of debugger / coverage tool / etc. It might break checks though, also it's only supported by CPython AFAIK, so proceed with caution.